### PR TITLE
[SPARK-45592][SQL] Correctness issue in AQE with InMemoryTableScanExec

### PR DESCRIPTION
**Pull request by eejbyfeldt**
### What changes were proposed in this pull request?
Fixes correctness issue in 3.5.0. The problem seems to be that when AQEShuffleRead does a coalesced read it can return a HashPartitioning with the coalesced number of partitions. This causes a correctness bug as the partitioning is not compatible for joins with other HashPartitioning even though the number of partitions matches. This is resolved in this patch by introducing CoalescedHashPartitioning and making AQEShuffleRead return that instead.

The fix was suggested by @cloud-fan 

> AQEShuffleRead should probably return a different partitioning, e.g. CoalescedHashPartitioning. It still satisfies ClusterDistribution, so Aggregate is fine and there will be no shuffle. For joins, two CoalescedHashPartitionings are compatible if they have the same original partition number and coalesce boundaries, and CoalescedHashPartitioning is not compatible with HashPartitioning.

### Why are the changes needed?
Correctness bug.

### Does this PR introduce _any_ user-facing change?
Yes, fixed correctness issue.

### How was this patch tested?
New and existing unit test.

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by eejbyfeldt** [comment]
Tagging @cloud-fan and @ulysses-you since they created PRs in this area and might not a better way of fixing the bug.

**Comment by ulysses-you** [comment]
I think the issue is that, we propagate a coalesced shuffle exchange through `InMemoryTableScanExec`, and then the `EnsureRequirements` use the coalesced shuffle exchange to create other side shuffle exchange.
However, the shuffle exchanges are actually not compatible. i.e., One side shuffle is from `HashPartitioning(200)` and then coalesce to  `HashPartitioning(10)` and other side shuffle is `HashPartitioning(10)`. So it causes the join data issue.
```
                      Scan
                       |
                   Shuffle(200)
                       |
  Scan           AQEShuffleRead(10)
   |                   |
Shuffle(10)   InMemoryTableScanExec
    \            /
         Join    
```

BTW, if you set `spark.sql.shuffle.partitions=5` , I think this issue should be resolved.

There are two code place related to this issue:
1. `AQEShuffleRead` always think the coalesced partitioning is not changed, so just refresh the partition number. I think it is based on the assumption that all the initial shuffle partition numbers are same but it seems not. The `EnsureRequirements` support `shouldConsiderMinParallelism` which cause different initial shuffle partition number in one query execution.
2. The `InMemoryTableScanExec` propagates the output partitioning. `InMemoryTableScanExec` would introduce one more query execution which also breaks the assumption of `AQEShuffleRead`

**Comment by cloud-fan** [comment]
`AQEShuffleRead` should probably return a different partitioning, e.g. `CoalescedHashPartitioning`. It still satisfies `ClusterDistribution`, so `Aggregate` is fine and there will be no shuffle. For joins, two `CoalescedHashPartitioning`s are compatible if they have the same original partition number and coalesce boundaries, and `CoalescedHashPartitioning` is not compatible with `HashPartitioning`.

**Comment by eejbyfeldt** [comment]
@cloud-fan I too a stab at implementing your suggestion, but the reproduction of the bug still fails. So either I made some mistake or missed some other part of the code that needs to be updated. Would be great if you could provide some feedback.

**Comment by cloud-fan** [comment]
@ulysses-you can you take a look when you have time?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 1]
```diff
@@ -700,14 +723,38 @@ case class HashShuffleSpec(
...
+    case CoalescedHashShuffleSpec(otherParent, otherPartitions) =>
+      partitions == otherPartitions &&
+      from.isCompatibleWith(otherParent)
+    case ShuffleSpecCollection(specs) =>
+      specs.exists(isCompatibleWith)
+    case _ =>
+      false
+  }
+
+  override def canCreatePartitioning: Boolean = from.canCreatePartitioning
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:750]: it can not create partitioning as it's not a standard hash partition. Otherwise `EnsureRequirememts` will take the created partitioning as the candidates partitioning.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 2]
```diff
@@ -700,14 +723,38 @@ case class HashShuffleSpec(
...
+
+  override def createPartitioning(clustering: Seq[Expression]): Partitioning =
+    CoalescedHashPartitioning(from.createPartitioning(clustering), partitions)
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:753]: So we do not need override `createPartitioning` since we can not create partitioning.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 3]
```diff
@@ -295,19 +284,53 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
...
     newChildren: IndexedSeq[Expression]): HashPartitioning = copy(expressions = newChildren)
 }
 
+case class CoalescedBoundary(startReducerIndex: Int, endReducerIndex: Int)
+
+/**
+ * Represents a partitioning where partitions have been coalesced from a HashPartitioning into a
+ * fewer number of partitions.
+ */
+case class CoalescedHashPartitioning(from: HashPartitioning, partitions: Seq[CoalescedBoundary])
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:320]: We'd better to avoid show the details of `partitions` for explain and UI.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 4]
```diff
@@ -700,14 +723,38 @@ case class HashShuffleSpec(
     }
   }
 
-  override def createPartitioning(clustering: Seq[Expression]): Partitioning = {
+  override def createPartitioning(clustering: Seq[Expression]): HashPartitioning = {
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:726]: Why do we need this change ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 1 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: That make sense, removed this and that made the spec reproducing the bug pass.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 3 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:318]: I added an override for `stringArgs` Is that the desired way to address this?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 4 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: Not needed, just me being sloppy when trying out some other approach. reverted.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 2 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala]: Removed.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 5]
```diff
@@ -295,7 +295,7 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
     }
   }
 
-  override def createShuffleSpec(distribution: ClusteredDistribution): ShuffleSpec =
+  override def createShuffleSpec(distribution: ClusteredDistribution): HashShuffleSpec =
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:298]: unnecessary change

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 6]
```diff
@@ -71,6 +71,19 @@ class DatasetSuite extends QueryTest
 
   private implicit val ordering = Ordering.by((c: ClassData) => c.a -> c.b)
 
+  test("SPARK-45592: correctness issue") {
```
Comment (ulysses-you) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala:74]: please refine the suite name

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 7]
```diff
@@ -306,6 +306,34 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
...
+
+}
+
+case class CoalescedBoundary(startReducerIndex: Int, endReducerIndex: Int)
+
+/**
+ * Represents a partitioning where partitions have been coalesced from a HashPartitioning into a
+ * fewer number of partitions.
+ */
+case class CoalescedHashPartitioning(from: HashPartitioning, partitions: Seq[CoalescedBoundary])
```
Comment (ulysses-you) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:318]: shall we add `@transient` for `partitions` ?

**Review by ulysses-you — COMMENTED** [review]
Looks fine to me, cc @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 5 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:298]: The refined type is used in `createShuffleSpec` to have more specific type in CoalescedHashShuffleSpec, but it not strictly needed so I will revert this change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 7 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:318]: Probably. What is the reason behind it? Is that we never intentionally what to serialize this class and if we do we will not really use it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 8]
```diff
@@ -306,6 +306,34 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
...
+  override def createShuffleSpec(distribution: ClusteredDistribution): ShuffleSpec =
+    CoalescedHashShuffleSpec(from.createShuffleSpec(distribution), partitions)
+
+  override protected def withNewChildrenInternal(
+    newChildren: IndexedSeq[Expression]): CoalescedHashPartitioning =
+      copy(from = from.copy(expressions = newChildren))
+
+  override val numPartitions: Int = partitions.length
+
+  override def stringArgs: Iterator[Any] = Iterator(from)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:336]: this looks weird, maybe just overwrite `toString` and `sql` to call `from.toString/sql`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 9]
```diff
@@ -708,6 +736,26 @@ case class HashShuffleSpec(
   override def numPartitions: Int = partitioning.numPartitions
 }
 
+case class CoalescedHashShuffleSpec(
+    from: ShuffleSpec,
+    @transient partitions: Seq[CoalescedBoundary]) extends ShuffleSpec {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:741]: what's the rationale of `@transient` here?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/DistributionSuite.scala [hunk 10]
```diff
@@ -146,63 +146,75 @@ class DistributionSuite extends SparkFunSuite {
...
-      ClusteredDistribution(Seq($"a", $"b", $"c"), requireAllClusterKeys = true),
-      false)
-
-    // HashPartitioning cannot satisfy OrderedDistribution
-    checkSatisfied(
-      HashPartitioning(Seq($"a", $"b", $"c"), 10),
-      OrderedDistribution(Seq($"a".asc, $"b".asc, $"c".asc)),
-      false)
+  private def testHashPartitioningLike(
+    partitioningName: String,
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/DistributionSuite.scala:150]: nit: 4 spaces indentation

**Comment by cloud-fan** [comment]
great work! With this change, we can do some cleanup in `ShuffledJoin`:
```
  override def requiredChildDistribution: Seq[Distribution] = {
    if (isSkewJoin) {
      // We re-arrange the shuffle partitions to deal with skew join, and the new children
      // partitioning doesn't satisfy `HashClusteredDistribution`.
      UnspecifiedDistribution :: UnspecifiedDistribution :: Nil
    } else {
      ClusteredDistribution(leftKeys) :: ClusteredDistribution(rightKeys) :: Nil
    }
  }
```
We don't need this hack anymore and can safely remove the `if` branch.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 9 above)
Comment (eejbyfeldt) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:741]: I added because of the suggestion of @ulysses-you here https://github.com/apache/spark/pull/43435#discussion_r1368494061

**Comment by eejbyfeldt** [comment]
> We don't need this hack anymore and can safely remove the `if` branch.

Is the suggestion to do that in this PR or is it better to do it in a follow up?

**Comment by cloud-fan** [comment]
@eejbyfeldt nvm, I made a mistake. This is for coalesce, we can add a new partitioning for skew join handling (split and replicate partitions). It's unrelated to this PR and we can do it latter.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala (hunk 9 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:741]: It is to avoid unnecessary ser/de. We only use it to check if compatible with other shuffle spec at driver side.

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:741]: which plan may hold `ShuffleSpec`? and do we need to do the same for other `ShuffleSpec`s?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:741]: I thought it would be holden at some place... After some search, it seems ShuffleSpec is created dynamically so it's unnecessary to consider ser/de. We can remove it @transient.

**Review by cloud-fan — APPROVED** [review]
LGTM with only one minor comment

**Comment by maryannxue** [comment]
@eejbyfeldt Can you briefly describe the triggering condition of this bug? Does it only occur when coalescing happens to produce just the exact number of partitions as the other side of the join?

In the meantime, I'm wondering if it would be better to:
1. not coalesce for the top/last shuffle of the physical plan of InMemoryTableScan
2. have coalesce rule deal with `InMemoryTableScan` from the caller side (user of the cache)

This PR, just to address the correctness issue, only needs to do 1. And we can do 2 (a little trickier I suppose) for performance improvement.

**Comment by maryannxue** [comment]
Synced with @cloud-fan offline, (2) in the above suggestion wouldn't work. Let's go ahead with current fix.

**Comment by cloud-fan** [comment]
The failed streaming test is unrelated, and my last comment is quite minor, let's merge it first to fix the correctness bug. Thanks for you great work!

**Comment by dongjoon-hyun** [comment]
Thank you, @eejbyfeldt and all.

**Comment by JJACOB0806** [comment]
Hello, is there a timeline for 3.5.1 release? We are facing the issue in 3.5.0 and would like to know when the next stable version will be rolled out.

**Comment by deepakcv** [comment]
Hi, is there a tentative timeline for releasing spark-3.5.1 with this release?
